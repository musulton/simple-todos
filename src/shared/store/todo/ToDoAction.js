import {ADD_TODO, CHANGE_TYPE, DELETE_TODO, SET_TODO_NAME, SHOW_LOADING, TOGGLE_COMPLETE} from "../../utils/constants";

export function addTodo(newTodo) {
    return {
        type: ADD_TODO,
        payload: newTodo
    }
}

export function toggleComplete(todoIndex) {
    return {
        type: TOGGLE_COMPLETE,
        payload: todoIndex
    }
}

export function deleteTodo(todoIndex) {
    return {
        type: DELETE_TODO,
        payload: todoIndex
    }
}

export function changeType(type) {
    return {
        type: CHANGE_TYPE,
        payload: type
    }
}
export function setTodoName(text) {
    return {
        type: SET_TODO_NAME,
        payload: text
    }
}

export function showLoading(loading) {
    return {
        type: SHOW_LOADING,
        payload: loading
    }
}
