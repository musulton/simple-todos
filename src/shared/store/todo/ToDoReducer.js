import {ADD_TODO, CHANGE_TYPE, DELETE_TODO, SET_TODO_NAME, SHOW_LOADING, TOGGLE_COMPLETE} from "../../utils/constants";

const initialState = {
    isLoading:false,
    newTodoName:'',
    todoIndex: 1,
    todo: [],
    type: 'All'
};
const ToDoReducer = (state = initialState, action) => {
    switch (action.type) {
        case ADD_TODO: {
            return {
                ...state,
                newTodoName: '',
                todoIndex: state.todoIndex + 1,
                todo: [...state.todo, action.payload]
            };
        }
        case TOGGLE_COMPLETE: {
            let todos = [...state.todo];
            todos.forEach((todo) => {
                if (todo.todoIndex === action.payload) {
                    todo.complete = !todo.complete;
                }
            })
            return {
                ...state,
                todo: todos
            };
        }
        case DELETE_TODO: {
            let todos = state.todo.filter((todo) => todo.todoIndex !== action.payload)
            return {
                ...state,
                todo: todos
            };
        }
        case CHANGE_TYPE: {
            return {
                ...state,
                type: action.payload
            };
        }
        case SET_TODO_NAME: {
            return {
                ...state,
                newTodoName: action.payload
            };
        }
        case SHOW_LOADING: {
            return {
                ...state,
                isLoading: action.payload
            };
        }
        default:
            return state;
    }
}
export default ToDoReducer;
