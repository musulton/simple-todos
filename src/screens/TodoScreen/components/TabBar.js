import React from 'react'
import {View, StyleSheet} from 'react-native'
import {useDispatch, useSelector} from "react-redux";

import {changeType} from "../../../shared/store/todo/ToDoAction";
import TabBarItem from "../../../shared/components/TabBarItem";

const TabBar = () => {
    const dispatch = useDispatch();
    const todoType = useSelector(state => state.ToDoReducer.type);

    const onSetType = (title) => {
        dispatch(changeType(title));
    }

    return (
        <View style={styles.container}>
            <TabBarItem border selected title='All' setType={onSetType} type={todoType}/>
            <TabBarItem border title='Active' setType={onSetType} type={todoType}/>
            <TabBarItem border title='Complete' setType={onSetType} type={todoType}/>
        </View>)
}

const styles = StyleSheet.create({
    container: {
        height: 70,
        flexDirection: 'row',
        borderTopWidth: 1,
        borderTopColor: '#dddddd'
    }
})
export default TabBar;
