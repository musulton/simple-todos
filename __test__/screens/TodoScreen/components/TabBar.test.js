import {fireEvent, render, screen} from "@testing-library/react-native";

import TabBar from "../../../../src/screens/TodoScreen/components/TabBar";
import {changeType} from "../../../../src/shared/store/todo/ToDoAction";

jest
    .mock("react-redux", () => ({
        ...jest.requireActual("react-redux"),
        useDispatch: jest.fn(() => jest.fn()),
        useSelector: jest.fn(() => "Active")
    }))
    .mock("../../../../src/shared/store/todo/ToDoAction", () => ({
        changeType: jest.fn()
    }))

describe("TabBar", () => {
  test("should return correctly", () => {
      render(<TabBar />)

      expect(screen.toJSON()).toMatchSnapshot()
  })

    test("should call changeType action when tab bar item is clicked", () => {
        render(<TabBar />)
        fireEvent.press(screen.getByText("Active"))

        expect(changeType).toHaveBeenCalledWith("Active")
    })
})
