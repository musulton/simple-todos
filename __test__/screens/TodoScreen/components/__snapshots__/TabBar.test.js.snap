// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabBar should return correctly 1`] = `
<View
  style={
    {
      "borderTopColor": "#dddddd",
      "borderTopWidth": 1,
      "flexDirection": "row",
      "height": 70,
    }
  }
>
  <View
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        },
        null,
        {
          "borderLeftColor": "#dddddd",
          "borderLeftWidth": 1,
        },
        null,
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#777777",
            "fontSize": 16,
          },
          null,
        ]
      }
    >
      All
    </Text>
  </View>
  <View
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        },
        null,
        {
          "borderLeftColor": "#dddddd",
          "borderLeftWidth": 1,
        },
        {
          "backgroundColor": "#ffffff",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#777777",
            "fontSize": 16,
          },
          {
            "fontWeight": "bold",
          },
        ]
      }
    >
      Active
    </Text>
  </View>
  <View
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        },
        null,
        {
          "borderLeftColor": "#dddddd",
          "borderLeftWidth": 1,
        },
        null,
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#777777",
            "fontSize": 16,
          },
          null,
        ]
      }
    >
      Complete
    </Text>
  </View>
</View>
`;
